libgalois/src/analytics/cdlp/cdlp.cpp

```diff
@@ -0,0 +1,302 @@
+              histogram[neighbor_data]++;
+            }
+
+            // Pick the most frequent communtiy as the new community for node
+            // pick the smallest one if more than one max frequent exist.
+            auto ndata_new_comm = ndata_current_comm;
+            size_t best_freq = 0;
+            for (const auto& [comm, freq] : histogram) {
+              if (freq > best_freq ||
+                  (freq == best_freq && comm < ndata_new_comm)) {
+                ndata_new_comm = comm;
+                best_freq = freq;
+              }
+            }
+
+            if (ndata_new_comm != ndata_current_comm)
+              applyBag.push(NodeDataPair(node, (CommunityType)ndata_new_comm));
+          },
+          katana::loopname("CDLP_Gather"));
+
+      // No change! break!
+      if (applyBag.empty())
+        break;
+
+      // Apply Phase
+      katana::do_all(
+          katana::iterate(applyBag),
+          [&](const NodeDataPair nodeData) {
+            GNode node = nodeData.node;
+            auto& ndata = graph->GetData<NodeCommunity>(node);
+            ndata = nodeData.data;
+          },
+          katana::loopname("CDLP_Apply"));
+
+      applyBag.clear();
+      iterations += 1;
+    }
+    katana::ReportStatSingle("CDLP_Synchronous", "iterations", iterations);
+  }
+};
+
+struct CdlpAsynchronousAlgo {
+  using CommunityType = uint64_t;
+  struct NodeCommunity : public katana::PODProperty<CommunityType> {};
+
+  using NodeData = std::tuple<NodeCommunity>;
+  using EdgeData = std::tuple<>;
+  using Graph = katana::TypedPropertyGraphView<
+      katana::PropertyGraphViews::Undirected, NodeData, EdgeData>;
+  using GNode = typename Graph::Node;
+
+  CdlpPlan& plan_;
+  CdlpAsynchronousAlgo(CdlpPlan& plan) : plan_(plan) {}
+
+  void Initialize(Graph* graph) {
+    katana::do_all(katana::iterate(*graph), [&](const GNode& node) {
+      auto& ndata = graph->GetData<NodeCommunity>(node);
+      ndata = node;
+    });
+  }
+
+  void Deallocate(Graph*) {}
+
+  void operator()(Graph*, size_t) {}
+};
+
+}  //namespace
+
+template <typename Algorithm>
+static katana::Result<void>
+CdlpWithWrap(
+    katana::PropertyGraph* pg, std::string output_property_name, CdlpPlan plan,
+    size_t max_iterations) {
+  katana::EnsurePreallocated(
+      2,
+      pg->topology().num_nodes() * sizeof(typename Algorithm::NodeCommunity));
+  katana::ReportPageAllocGuard page_alloc;
+
+  if (auto r = ConstructNodeProperties<
+          std::tuple<typename Algorithm::NodeCommunity>>(
+          pg, {output_property_name});
+      !r) {
+    return r.error();
+  }
+  auto pg_result = Algorithm::Graph::Make(pg, {output_property_name}, {});
+  if (!pg_result) {
+    return pg_result.error();
+  }
+  auto graph = pg_result.value();
+
+  Algorithm algo(plan);
+
+  algo.Initialize(&graph);
+
+  katana::StatTimer execTime("CDLP");
+
+  execTime.start();
+  algo(&graph, max_iterations);
+  execTime.stop();
+
+  algo.Deallocate(&graph);
```

Comment:
I don't think deallocate is required in this case. It was introduced for some cc algorithms. Please remove this.